### ISSUE TITLE
Implement css customisation options for themeing the plugin

### ISSUE THREAD

**Issue by kontur** [kontur/fontsampler-wordpress-plugin#7]
As a first level it would be great to be able to change the appearance of the plugin by some simple settings, like:

* changing main colour variables
* changing removing / changing layout borders 

It might further be good to implement at least some class or id for each font-sampler so people can manually overwrite styles for specific font samplers as well.

Down the line it would be great if those could be done even via the admin interface as additional styles.

**Comment by drawcard** [comment]
This is something I can take a look at. Should be pretty straightforward to set up the WP colour picker function to specify colours etc. Might also be a nice idea to have an advanced setting for the user to add their own CSS overrides inside the plugin.

**Comment by kontur** [comment]
In terms of interface I was thinking this could be located in the Settings tab as a new section, and from there the user can set customisations to all Fontsampler instances on the site.

While these values could be maintained in the database, I was wondering if it is best for performance to generate a css file from them and serve that in the actual front end instead of reading and generating the statements from the database each page view.

Adding custom css as a field either in the settings tab or then on a per Fontsampler instance could be a separated and possibly more trivial to implement, but less user friendly, option. The two are not exclusive, but I thin the color settings / customisation should take priority. 

Let me know how things progress.

**Comment by drawcard** [comment]
I had a bit of time up my sleeve today and I made a bit of headway with the setup - these 2 colour  values get written to a couple of new columns in the 'settings' table:

![image](https://cloud.githubusercontent.com/assets/1435638/19304397/08293b98-90b4-11e6-9879-02ec1e40735c.png)

I now just need to wire it up to the front-end presentation.

If I get a bit more time I might also look into customizing the colours on an individual font-set basis, taking note of how you've handled the other settings too. 

I like your suggestion of writing the values directly into the CSS rather than storing it in the DB, I might look into that to. It might be worthwhile if we look into custom colour settings for each font set, which could increase the load time if there's dozens of sets on a single page.

You can see what I'm up to in this branch: https://github.com/drawcard/fontsampler-wordpress-plugin/tree/color-settings

**Comment by drawcard** [comment]
Early tests are looking good! Changes pushed to my branch, I'll submit a PR once I've done a little bit more spring cleaning.

![image](https://cloud.githubusercontent.com/assets/1435638/19310288/e386eb9c-90d0-11e6-9d0e-b3511c23e3c4.png)

![image](https://cloud.githubusercontent.com/assets/1435638/19310327/0aba2a6c-90d1-11e6-98d1-b96cd32f48be.png)

**Comment by kontur** [comment]
Great work so far, this helps a lot. It'll be easy to duplicate the same picker for other fields, say if we want to affect the UI text color or icon colors, hover states etc.. Also remember to include a copy of your sql schema in the PR. I'll have to look into a future proof way of updating the table structure also with life updates of the plugin, so this could be a first use case to set that up.

For the settings table I had the idea of eventually using the same table for per-instance settings, once implemented. The general settings, the only row right now, would then just have a "defaults" boolean checked and there can only be one. Where no instance specific settings (base on fontsampler id) exist, the defaults are used.

In regards to the css serving: What if changing any of the css settings will call a routine to generate a new css file. In that file we have the cloned contents of the out-of-the-box front end styles and appended to it the custom settings. Any change to the settings would recreate this css file and we can reference it from the fontsampler.php enqueues - this would mean only one css file gets served in total (This would further allow implementing a concatenation of the custom rangeslider and selectric styles). The default styles could also be generated from the db defaults then in the future.

In regards to future instance-specific styles there could be two approaches. Same as before, just also append all css-prefixed (say with id of the fontsampler) overwrites of all instances compiled into the same file. Or then serve additional css files on the pages where instances are present - but resulting in several file requests.

**Comment by kontur** [comment]
Great stuff! Ideally you could provide me with a sql statement / file that alters the table in such a way, that running the sql will update the table setup I have to the table setup you have. There is ways to create these database migrations automatically, but I think a simple `ALTER TABLE ... ADD COLUMN...` or the like is a reasonable approach.

**Comment by drawcard** [comment]
No worries. Actually if you check out the edits I made to fontsampler.php you may find the SQL statements you're looking for, to create two new columns in MySQL: https://github.com/drawcard/fontsampler-wordpress-plugin/commit/fdd508cab8b7b9c5cb24ea335c245db1c5d49ef9 (line 346 & 347 in particular)

After adding these, I just refreshed the FontSampler plugin page in my browser and the create_table_settings() function was run, creating the necessary tables. 

If you need something more than that, let me know :)

**Comment by kontur** [comment]
I pulled in your changes to the dev branch, since I was expecting quite some merge conflicts due to some coding style changes I made yesterday and didn't want you to have to figure them out. At the same time I adjusted some of your code to integrate better, to the best of my abilities.

On the dev branch I further tested a routine that can be used to migrate the database from one version to the next, so this was a good use case to test it.

Right now one of the issues to work around is how you apply the styles, since both the `span` applying foreground color inside the textfield will get overwritten (when selecting text dummy presets like 'hamburgerfontstiv', since they replace the text inside the textfield) as well as the background color applied via a style attribute (when the jquery plugin writes new opentype features to the style attribute, rather bluntly). I've added that each instance now has `class="fontsampler fontsampler-id-xxx"`. Until we implement the setup that writes and merges different css files for all customisations you could simply output a `<style>` block just before the fontsampler instance gets printed to the page and make your color overwrites there, for example, so they don't get overwritten on DOM updates. For now you can even just overwrite the colours with the .fontsampler class since there is no per-instance settings.

To merge in the dev branch just `git pull origin dev` it with your current `color-options` branch active and you should be prompted to merge. The merge should fast forward automatically, since I pulled in all your latest changes, unless you have new unpublished commits. Let me know if you need help with that. 👍

**Comment by drawcard** [comment]
Thanks for that - I will have a play with it over the weekend!

**Comment by kontur** [comment]
Note when pulling in the dev branch that I've moved the outputting of the `<div class="fontsampler">...` to the file `includes/interface.php` since the position of all UI elements is now customisable in the back end.

You'll further need to create one additional field in your wp_fontsampler_sets table (given your wordpress tableprefix is indeed wp_): 
`ALTER TABLE wp_fontsampler_sets ADD ui_order VARCHAR( 255 ) NOT NULL DEFAULT 'size,letterspacing,options|fontpicker,sampletexts,lineheight|fontsampler'`

**Comment by kontur** [comment]
Made another small database addition for the 0.0.2 release. Check the function `migrate_db` in the main file; it has all the necessary changes in one place.

**Comment by drawcard** [comment]
Cheers, that makes more sense having it in the interface file.

**Comment by kontur** [comment]
I've now implemented the css file merge as discussed above. The settings (I added several more) are now added to the settings table and rendered via a css template into a css file that is then embedded on the front end side of things.

There is sure more to polish this some more, but for now closing the issue as the basic functionality is there. Thanks again @drawcard for your help setting up the color pickers 👍 

This change is now also in master and bumps the version to 0.0.2
